File: leetcode/366.find-leaves-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {

private:
    unordered_map<int, vector<int>> heightMap;
    int nodeHeight(TreeNode* node)
    {
        if (node == NULL)
        {
            return 0;
        }

        int height = max(nodeHeight(node->left) + 1, nodeHeight(node->right) + 1);
        heightMap[height].push_back(node->val);

        return height;
    }

public:
    vector<vector<int>> findLeaves(TreeNode* root) {
        int max_height = nodeHeight(root);

        vector<vector<int>> res;
        for(auto i = 1; i <= max_height; ++i)
        {
            res.push_back(heightMap[i]);
        }

        return res;
    }
};

File: leetcode/366.find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self, *args, **kwargs):
        self.heightMap = {}

    def node_height(self, cur):
        if cur is None:
            return 0

        height = max(self.node_height(cur.left) + 1, self.node_height(cur.right) + 1)
        if height in self.heightMap:
            self.heightMap[height].append(cur.val)
        else:
            self.heightMap[height] = [cur.val]

        return height

    def findLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        max_height = self.node_height(root)
        res = []

        for i in range(1, max_height + 1):
            res.append(self.heightMap[i])

        return res
        
